app/actions/client.py

```diff
@@ -0,0 +1,162 @@
+                           'ac.speed,ac.hdop,ac.altitude,ac.heading,u.description'
+
+
+class GalooliGeneralErrorException(Exception):
+    def __init__(self, error: Exception, message: str, code=-1):
+        self.code = code
+        self.message = message
+        self.error = error
+        super().__init__(f"'{self.code}: {self.message}, Error: {self.error}'")
+
+
+class GalooliInvalidUserCredentialsException(Exception):
+    def __init__(self, error: Exception, message: str, code=1000):
+        self.code = code
+        self.message = message
+        self.error = error
+        super().__init__(f"'{self.code}: {self.message}, Error: {self.error}'")
+
+
+class GalooliTooManyRequestsException(Exception):
+    def __init__(self, error: Exception, message: str, code=1101):
+        self.code = code
+        self.message = message
+        self.error = error
+        super().__init__(f"'{self.code}: {self.message}, Error: {self.error}'")
+
+
+def convert_to_er_observation(r, reports_timezone):
+    sensor_id = r[0]
+    subject_name = r[1]
+    asset_model = r[2]
+    org_name = r[3]
+    time = r[5]
+    status = r[6]
+    latitude = r[7]
+    longitude = r[8]
+    distance = r[9]
+    speed = r[10]
+    hdop = r[11]
+    altitude = r[12]
+    heading = r[13]
+    description = r[14]
+
+    if latitude and longitude and time and sensor_id:
+        if status == 'Moving':
+            localized_gps_time = reports_timezone.localize(dateparser.parse(time))
+            obs = {
+                'manufacturer_id': sensor_id,
+                'subject_name': subject_name,
+                'subject_groups': ['Vehicles', ],
+                'subject_subtype': "security_vehicle",
+                'recorded_at': localized_gps_time.isoformat(),
+                'location': {
+                    'lat': latitude,
+                    'lon': longitude
+                },
+                'additional': {
+                    'sensor_id': sensor_id,
+                    'asset_model': asset_model,
+                    'org_name': org_name,
+                    'status': status,
+                    'distance': distance,
+                    'speed': speed,
+                    'hdop': hdop,
+                    'altitude': altitude,
+                    'heading': heading,
+                    'description': description,
+                }
+            }
+            return obs
+        else:
+            logger.debug(f'Ignoring observation for {sensor_id} {subject_name} recorded at {time} status: {status}')
+            return None
+    else:
+        logger.error(f'Got bad data from Galooli: mfg_id {sensor_id}, time {time}, lat {latitude} long {longitude}')
+        logger.error(f'{r}')
+        return None
+
+
+@stamina.retry(on=GalooliTooManyRequestsException, wait_initial=4.0, wait_jitter=5.0, wait_max=32.0)
+async def get_observations(integration, url, auth_config, pull_config):
+    async with httpx.AsyncClient(timeout=120) as session:
```

Comment:
[nitpick] I would prefer this to not take Gundi objects as parameters but instead take the essential data that's relevant tot he Galooli API. Keeping things separate will help keep this coherent and resilient to future changes.